File: Rust/2023/day1/solution.rs
use std::env;
use std::fs;

fn main() {
    let args: Vec<String> = env::args().collect();

    let input = fs::read_to_string(&args[1]).expect("Read input correctly");
    let lines = input.lines().collect::<Vec<_>>();

    println!("Part 1: {}", solve(&lines, false));
    println!("Part 1: {}", solve(&lines, true));
}

fn solve(lines: &[&str], part2: bool) -> u32 {
    let digit_names = [
        "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
    ];

    lines
        .iter()
        .map(|l| {
            let mut digits = vec![];
            for (i, c) in l.char_indices() {
                match c.to_digit(10) {
                    Some(x) => digits.push(x),
                    None => {
                        if part2 {
                            match digit_names.iter().position(|s| l[i..].starts_with(s)) {
                                Some(pos) => digits.push((pos + 1) as u32),
                                None => (),
                            }
                        }
                    }
                }
            }
            digits[0] * 10 + digits[digits.len() - 1]
        })
        .sum()
}

File: JS/2023/day1/solution.js
const fs = require("node:fs");

let lines = fs.readFileSync("input.txt", "utf8").split("\n");
let numbers = {
    one: 1,
    two: 2,
    three: 3,
    four: 4,
    five: 5,
    six: 6,
    seven: 7,
    eight: 8,
    nine: 9,
};

function solve(lines, part2) {
    return lines
        .map((l) => {
            let digits = [];
            for (const [i, c] of l.split("").entries()) {
                if (c >= "0" && c <= "9") digits.push(c - "0");
                else if (part2) {
                    for (const size of [3, 4, 5]) {
                        let chunk = l.substr(i, size);
                        if (chunk in numbers) digits.push(numbers[chunk]);
                    }
                }
            }

            return digits.at(0) * 10 + digits.at(-1);
        })
        .reduce((acc, v) => acc + v, 0);
}

console.log("Part 1: %d", solve(lines, false));
console.log("Part 2: %d", solve(lines, true));
